addons/udes_sale_stock/models/stock_picking.py

```diff
@@ -0,0 +1,20 @@
+from odoo import api, fields, models, _
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    def action_cancel(self):
+        """
+        Set is_cancelled on relevant sale_lines when cancelling a picking
+        """
+        location_customers = self.env.ref('stock.stock_location_customers')
+        res = super(StockPicking, self).action_cancel()
+        for move in self.mapped('move_lines'):
+            sale_line = move.sale_line_id
+            if sale_line and not sale_line.is_cancelled:
+                not_cancelled = sale_line.move_ids.filtered(
+                    lambda m: m.state not in ['cancel'] and
+                              m.location_dest_id == location_customers)
+                sale_line.is_cancelled = len(not_cancelled) == 0
```

Comment:
Batch all the sale lines to set is_cancelled = True and do it in one call .write()

Reply:
changed this a bit, looks mental but works .. using ORM API calls instead of loops